File: MatlabBlenderIO/csvtext2matrix.py
import numpy as np
import re
from itertools import groupby

def csvtext2matrix(text):
    text = str(text)
    text = text.replace("\n", "")
    if not "," in text:
        mat = np.array(float(text))
        dimension_lengths = 0
    else:

        # Find highest and lowest delimiters:
        highest_delimiter_number, lowest_delimiter_number = find_comma_sequences(text)
        highest_delimiter = highest_delimiter_number*","
        lowest_delimiter  = lowest_delimiter_number *","

        
        # Find the length of each dimension:
        dimensions         = highest_delimiter_number - lowest_delimiter_number+1
        dimension_lengths  = [0]*dimensions
        dimension_iterator = dimensions
        delimiter_iterator = highest_delimiter
        module             = 1

        while True:
            data_split_along_dim                     = text.split(delimiter_iterator)
            data_split_along_dim                     = list(filter( lambda x: x!="", data_split_along_dim ))
            dimension_lengths[dimension_iterator-1]  = int(len(data_split_along_dim)/module)
            module                                  *= dimension_lengths[dimension_iterator-1]

            if delimiter_iterator == lowest_delimiter:
                break
            else:
                delimiter_iterator  = delimiter_iterator[:-1]
                dimension_iterator -= 1

        # Making the matrix:
        flat_data = text.split(",")
        flat_data = list(filter( lambda x: x!="", flat_data ))
        flat_data = list(map(float, flat_data ))

        
        mat = np.array(flat_data)
        if dimensions > 1:
            mat = np.reshape(mat, dimension_lengths, 'F')

    return mat, dimension_lengths



def find_comma_sequences(text):
    comma_sequences = re.findall(r',+', text)
    if comma_sequences:
        longest  = max(len(seq) for seq in comma_sequences)
        smallest = min(len(seq) for seq in comma_sequences)
        return longest, smallest
    else:
        return 0, 0





def csvtext2matrix_demo():

    print(csvtext2matrix("0,0,0,,0,0,0,,,0,0,0,,0,0,0,,,0,0,0,,0,0,0"))
    print(csvtext2matrix("0,0,0,,0,0,0,,,0,0,0,,0,0,0"))
    print(csvtext2matrix("0,0,0,,0,0,0"))
    print(csvtext2matrix("0,1,3"))
    print(csvtext2matrix("0"))
    print(csvtext2matrix("0,1,0,,0,0,0,,,0,5,0,,8,0,0"))
    print(csvtext2matrix("0,,,1,,,2,,,4"))
    print(csvtext2matrix("0,,,,1,,,,2,,,,4"))
    testmat, testdims = csvtext2matrix("0,1,0,,0,0,0,,,0,5,0,,8,0,0")
    print(testmat[1][0][0])
    print(testmat[1][0][1])
    print(testmat[0][1][1])
    print(csvtext2matrix("0,1,0,0,0,0,,,0,5,0,8,0,0"))



csvtext2matrix_demo()

File: MatlabBlenderIO/csvtext2matrix.m
function [mat, dimension_length] = csvtext2matrix(text)

text = erase(text, char(10));
text = string(text);
if contains(text, ",")

[highest_delimiter_number, lowest_delimiter_number] = find_comma_sequences(text);
highest_delimiter = strjoin(repmat(",", 1,highest_delimiter_number),"");
lowest_delimiter  = strjoin(repmat(",", 1,lowest_delimiter_number ),"");


dimensions         = highest_delimiter_number - lowest_delimiter_number+1;
dimension_length   = zeros(1,dimensions);
delimiter_iterator = highest_delimiter;
dimension_iterator = dimensions;
module             = 1;

while true
data_split_along_dim                 = split(text, delimiter_iterator);
data_split_along_dim                 = data_split_along_dim(~strcmp(data_split_along_dim, ""));
dimension_length(dimension_iterator) = numel(data_split_along_dim)/module;
module                               = module*dimension_length(dimension_iterator);

if isequal(delimiter_iterator, lowest_delimiter)
break
else
delimiter_iterator = extractBefore(delimiter_iterator, strlength(delimiter_iterator));
dimension_iterator = dimension_iterator -1;
end

end

flat_data          = split(text, ",");
flat_data          = flat_data(~strcmp(flat_data, ""));
flat_data          = double(flat_data);
if numel(dimension_length)>1; mat = reshape(flat_data, dimension_length);
else;                         mat = flat_data;
end

else
mat = double(text); dimension_length = 0;


end


end



function [longest, smallest] = find_comma_sequences(text)

comma_sequences = regexp(text, ',+', 'match');
lengths         = cellfun(@length, comma_sequences);
    
if ~isempty(lengths); longest = max(lengths); smallest = min(lengths);
else;                 longest = 0;            smallest = 0;
end

end